File: Day06/Part02/day06_02.java
import java.io.BufferedReader;
import java.io.FileReader;
import java.util.ArrayList;
import java.util.Arrays;


class day06_02 {


    public static void main(String[] args) {

        // get stacking blocks 
        Integer[] blocks = getBlocks("input.txt");
        System.out.print("Original Blocks: ");
        for (Integer i : blocks) {
            System.out.print(i + " ");
        }

        // create arraylist to hold previous block patterns
        ArrayList<String> blockPatterns = new ArrayList<String>();

        // keep rearranging while the current block pattern has not been seen
        Integer loopCount = 0;
        while (!blockPatterns.contains(Arrays.toString(blocks))) {
            // add current blocks pattern to the arraylist
            blockPatterns.add(Arrays.toString(blocks));
            // cycle the blocks
            blocks = cycleBlocks(blocks);
            // increment loop counter
            loopCount++;
        }
        
        System.out.println("\nLoop Count: " + loopCount);

        // reset arraylist for part 2
        blockPatterns = new ArrayList<String>();

        Integer cycleCount = 0;

        while (!blockPatterns.contains(Arrays.toString(blocks))) {
            // add current blocks pattern to the arraylist
            blockPatterns.add(Arrays.toString(blocks));
            // cycle the blocks
            blocks = cycleBlocks(blocks);
            // increment loop counter
            cycleCount++;
        }

        System.out.println("Cycle Count: " + cycleCount);

    }


    public static Integer[] getBlocks(String filename) {

        Integer[] blocks = null;

        try {  

            BufferedReader br = new BufferedReader(new FileReader(filename));
            String[] stringBlocks = br.readLine().split("\\s+");
            blocks = new Integer[stringBlocks.length];
            for (int i = 0; i < stringBlocks.length; i++) {
                blocks[i] = Integer.parseInt(stringBlocks[i]);
            }
            br.close();
            
        } catch (Exception e) {
            e.printStackTrace();
        }

        return blocks;

    }


    public static Integer[] cycleBlocks(Integer[] blocks) {

        Integer largestBlock = 0;
        Integer toRedistribute = null;

        // find largest block
        for (int i = 0; i < blocks.length; i++) {
            if (blocks[i] > blocks[largestBlock]) {
                largestBlock = i;
            }
        }

        // redistribute to other blocks
        toRedistribute = blocks[largestBlock];
        blocks[largestBlock] = 0;
        while (toRedistribute > 0) {
            largestBlock++;
            if (largestBlock >= blocks.length) {
                largestBlock = 0;
            }
            blocks[largestBlock]++;
            toRedistribute--;
        }

        return blocks;
    }

}

File: Day06/Part02/day06_02.py
def getBlocks(filename):
    file = open(filename, 'r')
    return [int(i) for i in file.read().split()]


if __name__ == '__main__':

    # get initial distribution of blocks
    blocks = getBlocks('input.txt')

    # keep track of past patterns
    past_blocks = []

    # keep redistributing blocks until a pattern repeats
    loops = 0
    while not (blocks in past_blocks):
        loops += 1
        # add copy of current distribution
        past_blocks.append(list(blocks))
        # find the largest block
        largest_block_position = 0
        largest_block_size = blocks[0]
        for i,s in enumerate(blocks):
            if (s > largest_block_size):
                largest_block_position = i
                largest_block_size = s
        # redistribute the block
        to_redistribute = largest_block_size
        position = largest_block_position
        blocks[position] = 0
        while (to_redistribute != 0):
            position += 1
            if position >= len(blocks):
                position = 0
            blocks[position] += 1
            to_redistribute -= 1

    # print number of loops before a repeat
    print("Number Of Loops Before First Repeat:", loops)

    # now keep track of number of loops in each cycle
    cycle_count = 0
    block_target = list(blocks)
    next_block = None
    while (block_target != next_block):
        # find the largest block
        largest_block_position = 0
        largest_block_size = blocks[0]
        for i,s in enumerate(blocks):
            if (s > largest_block_size):
                largest_block_position = i
                largest_block_size = s
        # redistribute the block
        to_redistribute = largest_block_size
        position = largest_block_position
        blocks[position] = 0
        while (to_redistribute != 0):
            position += 1
            if position >= len(blocks):
                position = 0
            blocks[position] += 1
            to_redistribute -= 1
        # set temp variable to check if it matches target
        next_block = list(blocks)
        cycle_count += 1
    
    # print how many cycles before repeat
    print("Number Of Loops In Final Cycle:", cycle_count)